mars/services/task/execution/ray/shuffle.py

```diff
@@ -0,0 +1,162 @@
+                reducer_subtasks, n_reducers
+            )
+            self.reducer_indices.update(
+                {
+                    subtask: (shuffle_index, reducer_ordinal)
+                    for reducer_ordinal, subtask in enumerate(
+                        sorted_filled_reducer_subtasks
+                    )
+                }
+            )
+
+    @staticmethod
+    def _get_sorted_filled_reducers(
+        reducer_subtasks: Iterable[Subtask], n_reducers: int
+    ):
+        # For operands such as `PSRSAlign`, sometimes `reducer_subtasks` might be less than `n_reducers`.
+        # fill missing reducers with `None`.
+        filled_reducers = list(range(n_reducers))
+        for subtask in reducer_subtasks:
+            reducer_ordinal = _get_reducer_operand(subtask.chunk_graph).reducer_ordinal
+            filled_reducers[reducer_ordinal] = subtask
+        return filled_reducers
+
+    def has_shuffle(self):
+        """
+        Whether current subtask graph has shuffles to execute.
+        """
+        return self.num_shuffles > 0
+
+    def add_mapper_output_refs(
+        self, subtask, output_object_refs: List["ray.ObjectRef"]
+    ):
+        """
+        Record mapper output ObjectRefs which will be used by reducers later.
+
+        Parameters
+        ----------
+        subtask
+        output_object_refs : List["ray.ObjectRef"]
+            Mapper output ObjectRefs.
+        """
+        shuffle_index, mapper_index = self.mapper_indices[subtask]
+        self.mapper_output_refs[shuffle_index][mapper_index] = np.array(
+            output_object_refs
+        )
+
+    def get_reducer_input_refs(self, subtask) -> List["ray.ObjectRef"]:
+        """
+        Get the reducer inputs ObjectRefs output by mappers.
+
+        Parameters
+        ----------
+        subtask : Subtask
+            A reducer subtask.
+        Returns
+        -------
+        input_refs : List["ray.ObjectRef"]
+            The reducer inputs ObjectRefs output by mappers.
+        """
+        shuffle_index, reducer_ordinal = self.reducer_indices[subtask]
+        return self.mapper_output_refs[shuffle_index][:, reducer_ordinal]
+
+    def get_n_reducers(self, subtask):
+        """
+        Get the number of shuffle blocks that a mapper operand outputs,
+        which is also the number of the reducers when tiling shuffle operands.
+        Note that this might be greater than actual number of the reducers in the subtask graph,
+        because some reducers may not be added to chunk graph.
+
+        Parameters
+        ----------
+        subtask : Subtask
+            A mapper or reducer subtask.
+        Returns
+        -------
+        n_reducers : int
+            The number of shuffle blocks that a mapper operand outputs.
+        """
+        mapper_index = self.mapper_indices.get(subtask)
+        if mapper_index:
+            shuffle_index = mapper_index[0]
+        else:
```

Comment:
When goes to `else`?

Reply:
If the `subtask` isn't a mapper sutbask, it will go to `else` branch.

Reply:
This is a bit clearer:
```python
index = self.mapper_indices.get(subtask) or self.reducer_indices.get(subtask)
if index is None:
    raise Exception("The {subtask} should be a mapper or a reducer.")
else:
    shuffle_index, _ = index
```

Reply:
I noticed that the `get_n_reducers` is called only when the subtask is mapper, so is the `else` needed?

Reply:
`else` can be reserved for use in future